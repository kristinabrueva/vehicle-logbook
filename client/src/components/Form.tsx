import React, { useState } from "react";
import { MODELS, POPULAR_CARS } from "../data";
import Button from "./Button";
import Select from "./Select";

const Form: React.FunctionComponent = () => {
  const data = MODELS;
  const popularCars = POPULAR_CARS;

  const [error, setError] = useState<boolean | undefined>(undefined);

  // selected values
  const [make, setMake] = useState<string>("");
  const [model, setModel] = useState<string>("");
  const [badge, setBadge] = useState<string>("");
  const [file, setFile] = useState<File>();

  // select options
  const makeOptions = Object.keys(data);

  const modelOptions = () => Object.keys(data[make as keyof typeof data]);

  const badgeOptions = () => {
    const dataModel = data[make as keyof typeof data];
    return dataModel[model as keyof typeof dataModel];
  };

  const handleMakeChange = (event: React.ChangeEvent): void => {
    setBadge("");
    setModel("");
    setMake((event.target as HTMLInputElement).value);
  };
  const handleModelChange = (event: React.ChangeEvent): void => {
    setBadge("");
    setModel((event.target as HTMLInputElement).value);
  };
  const handleBadgeChange = (event: React.ChangeEvent): void => {
    setBadge((event.target as HTMLInputElement).value);
  };

  const handleCommonVehicleSelect = (
    e: React.MouseEvent,
    mk: string,
    md: string,
    bg: string
  ) => {
    e.preventDefault();
    setMake(mk);
    setModel(md);
    setBadge(bg);
  };

  const handleSubmit = (event: React.FormEvent): void => {
    if (!file) {
      setError(true);
    } else {
      let data = new FormData();

      data.append("make", make);
      data.append("model", model);
      data.append("badge", badge);
      data.append("file", file);

      fetch("http://localhost:4000/upload", {
        method: "POST",
        body: data,
      }).then(function (response) {
        window.location.href = "http://localhost:4000/upload";
        return response.json();
      });
    }

    event.preventDefault();
  };
  return (
    <form
      className="flex flex-col text-left p-8 gap-6"
      key="form"
      onSubmit={handleSubmit}
    >
      <h1 className="font-bold text-6xl">Drill Down Form</h1>
      <div className="flex flex-col gap-1">
        <Select
          name="make"
          value={make}
          options={makeOptions}
          testId="testMakeSelect"
          handleChange={handleMakeChange}
        />
        {make && (
          <Select
            name="model"
            value={model}
            options={modelOptions()}
            testId="testModelSelect"
            handleChange={handleModelChange}
          />
        )}
        {model && (
          <Select
            name="badge"
            value={badge}
            options={badgeOptions()}
            testId="testBadgeSelect"
            handleChange={handleBadgeChange}
          />
        )}
      </div>
      {badge && (
        <div className="flex flex-col gap-5">
          <h3 className="text-lg font-medium">Upload Logbook</h3>
          <div className="flex flex-col gap-1">
            <input
              name="file"
              type="file"
              accept="text/plain"
              onChange={(e: React.ChangeEvent) => {
                const value = e.target as HTMLInputElement;
                if (error) setError(false);
                if (value.files) {
                  setFile(value.files[0]);
                }
              }}
              className={`text-sm w-fit line ${
                error && "after:content-['❗️'] after:pl-1"
              }
                  ${error === false && "after:content-['✅'] after:pl-1"}`}
            />
            {error && (
              <div className="text-rose-600 text-sm">Please upload a file</div>
            )}
            <Button type="submit" text="Submit"></Button>
          </div>
        </div>
      )}

      <h3 className="text-4xl font-bold">Select a vehicle</h3>
      <div className="flex flex-col gap-1">
        {popularCars.map((car, i) => (
          <Button
            id={i}
            key={i}
            text={Object.values(car).join(" ")}
            inline
            handleClick={(e) =>
              handleCommonVehicleSelect(e, car.make, car.badge, car.model)
            }
          />
        ))}
      </div>
    </form>
  );
};

export default Form;
