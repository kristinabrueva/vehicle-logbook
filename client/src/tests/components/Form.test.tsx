import React from "react";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Form from "../../components/Form";

describe("Tests Form", () => {
  it("should render default form with one select only", () => {
    const { container, getAllByRole, queryAllByText } = render(<Form />);
    expect(container).toMatchSnapshot();
    expect(getAllByRole("combobox")).toHaveLength(1);
    expect(queryAllByText("Submit")).toHaveLength(0);
  });

  it("simulates Submit with no file uploaded", async () => {
    const { getByTestId, queryAllByText, getAllByRole, getByRole, getByText } =
      render(<Form />);
    expect(queryAllByText("Submit")).toHaveLength(0);
    expect(getAllByRole("combobox")).toHaveLength(1);

    await userEvent.selectOptions(
      getByTestId("testMakeSelect"),
      getByRole("option", { name: "Tesla" })
    );

    expect(getAllByRole("combobox")).toHaveLength(2);

    await userEvent.selectOptions(
      getByTestId("testModelSelect"),
      getByRole("option", { name: "Model S" })
    );

    expect(getAllByRole("combobox")).toHaveLength(3);

    await userEvent.selectOptions(
      getByTestId("testBadgeSelect"),
      getByRole("option", { name: "Standart" })
    );

    expect(queryAllByText("Submit")).toHaveLength(1);

    await userEvent.click(getByText("Submit"));
    expect(getByText("Please upload a file")).toBeDefined();
  });
});
