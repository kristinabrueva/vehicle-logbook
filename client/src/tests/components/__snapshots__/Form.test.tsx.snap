// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests Form should render default form with one select only 1`] = `
<div>
  <form
    class="flex flex-col text-left p-8 gap-6"
  >
    <h1
      class="font-bold text-6xl"
    >
      Drill Down Form
    </h1>
    <div
      class="flex flex-col gap-1"
    >
      <select
        class="block border-2 max-w-[160px] text-left text-base"
        data-testid="testMakeSelect"
        id="make"
        name="make"
        required=""
      >
        <option
          disabled=""
          hidden=""
          value="default"
        >
          make
        </option>
        <option
          value="Ford"
        >
          Ford
        </option>
        <option
          value="bmw"
        >
          bmw
        </option>
        <option
          value="Tesla"
        >
          Tesla
        </option>
      </select>
    </div>
    <h3
      class="text-4xl font-bold"
    >
      Select a vehicle
    </h3>
    <div
      class="flex flex-col gap-1"
    >
      <button
        class="border-[1px] rounded-sm text-sm px-2 bg-gray-100 border-black py-1 max-w-fit inline"
        data-testid="testButton"
        id="0"
      >
        Tesla Model 3 Performance
      </button>
      <button
        class="border-[1px] rounded-sm text-sm px-2 bg-gray-100 border-black py-1 max-w-fit inline"
        data-testid="testButton"
        id="1"
      >
        bmw 130d xDrive 30d
      </button>
      <button
        class="border-[1px] rounded-sm text-sm px-2 bg-gray-100 border-black py-1 max-w-fit inline"
        data-testid="testButton"
        id="2"
      >
        Tesla Model S Long Range
      </button>
      <button
        class="border-[1px] rounded-sm text-sm px-2 bg-gray-100 border-black py-1 max-w-fit inline"
        data-testid="testButton"
        id="3"
      >
        Ford Ranger Raptor
      </button>
    </div>
  </form>
</div>
`;
